nomad/stream/event_buffer.go

```diff
@@ -0,0 +1,308 @@
+func (b *eventBuffer) Append(events structs.Events) {
+	b.appendItem(newBufferItem(events))
+}
+
+func (b *eventBuffer) appendItem(item *bufferItem) {
+	// Store the next item to the old tail
+	oldTail := b.Tail()
+	oldTail.link.next.Store(item)
+
+	// Update the tail to the new item
+	b.tail.Store(item)
+
+	// Increment the buffer size
+	size := atomic.AddInt64(b.size, 1)
+
+	// Check if we need to advance the head to keep the list
+	// constrained to max size
+	if size > b.maxSize {
+		b.advanceHead()
+	}
+
+	// notify waiters next event is available
+	close(oldTail.link.ch)
+
+}
+
+// advanceHead drops the current Head buffer item and notifies readers
+// that the item should be discarded by closing droppedCh.
+// Slow readers will prevent the old head from being GC'd until they
+// discard it.
+func (b *eventBuffer) advanceHead() {
+	old := b.Head()
+	next := old.link.next.Load()
+
+	close(old.link.droppedCh)
+	b.head.Store(next)
+	atomic.AddInt64(b.size, -1)
+
+}
+
+// Head returns the current head of the buffer. It will always exist but it may
+// be a "sentinel" empty item with a nil Events slice to allow consumers to
+// watch for the next update. Consumers should always check for empty Events and
+// treat them as no-ops. Will panic if eventBuffer was not initialized correctly
+// with NewEventBuffer
+func (b *eventBuffer) Head() *bufferItem {
+	return b.head.Load().(*bufferItem)
+}
+
+// Tail returns the current tail of the buffer. It will always exist but it may
+// be a "sentinel" empty item with a Nil Events slice to allow consumers to
+// watch for the next update. Consumers should always check for empty Events and
+// treat them as no-ops. Will panic if eventBuffer was not initialized correctly
+// with NewEventBuffer
+func (b *eventBuffer) Tail() *bufferItem {
+	return b.tail.Load().(*bufferItem)
+}
+
+// StarStartAtClosest returns the closest bufferItem to a requested starting
+// index as well as the offset between the requested index and returned one.
+func (b *eventBuffer) StartAtClosest(index uint64) (*bufferItem, int) {
+	item := b.Head()
+	if index < item.Index {
+		return item, int(item.Index) - int(index)
+	}
+	if item.Index == index {
+		return item, 0
+	}
+
+	for {
+		prev := item
+		item = item.NextNoBlock()
+		if item == nil {
+			return prev, int(index) - int(prev.Index)
+		}
+		if index < item.Index {
+			return item, int(item.Index) - int(index)
+		}
+		if index == item.Index {
+			return item, 0
+		}
+	}
+}
+
+// Len returns the current length of the buffer
+func (b *eventBuffer) Len() int {
+	return int(atomic.LoadInt64(b.size))
+}
+
+// prune advances the head of the buffer until the head buffer item TTL
+// is no longer expired. It should be externally synchronized as it mutates
+// the buffer of items.
+func (b *eventBuffer) prune() {
+	for {
+		head := b.Head()
+		if b.Len() == 0 {
+			return
+		}
+
+		if time.Since(head.createdAt) > b.maxItemTTL {
```

Comment:
May want to call `time.Now` once per prune. It may not prune as aggressively but will be cheaper when there are a large number of items to prune.